### PR TITLE
Configure indexer lambda with information about public snapshots (#3572)

### LINKED ISSUE #3572: Configure indexer lambda with information about public snapshots

### PR DESCRIPTION
**Pull request by nadove-ucsc**
https://github.com/DataBiosphere/azul/issues/3572

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [ ] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [ ] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [ ] Decided if PR can be labeled `no sandbox`
- [ ] PR title is appropriate as title of merge commit
- [ ] Moved ticket to Approved column
- [ ] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [ ] Checked `reindex` label and `r` commit title tag
- [ ] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [ ] Rebased and squashed branch
- [ ] Sanity-checked history
- [ ] Pushed PR branch to Github
- [ ] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [ ] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [ ] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Added PR reference to merge commit title
- [ ] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
Change in lambda configuration compared to develop:
Screenshot from 2021-11-19 16-08-08
Screenshot from 2021-11-19 16-08-22

**Review by amarjandu — CHANGES_REQUESTED** [review]
I think some of the logic with things like `catalog.replace` and the actual value of the env var key should be contained within the `Config.Catalog` class.


On a side note/nit, I think its weird that we expose `config.sources` as a method, why not just access the `config.catalogs[name].sources` directly? I like the idea of encapsulating all the catalog logic within the `Config.Catalog` class, `managed_access_source_ids` would fit in nicely as a property.

**Comment by noah-aviel-dove** [comment]
> I think some of the logic with things like catalog.replace and the actual value of the env var key should be contained within the Config.Catalog class.

I don't think this is feasible because the value needs to be plugin-specific.

> On a side note/nit, I think its weird that we expose config.sources as a method, why not just access the config.catalogs[name].sources directly? I like the idea of encapsulating all the catalog logic within the Config.Catalog class, managed_access_source_ids would fit in nicely as a property.

Interesting idea. I'll bring it up at parking lot.

**Comment by noah-aviel-dove** [comment]
We'll have two classes: `TDRSnapshot` to represent the snapshots as they exist in TDR without any of our configuration, with fields `id`, `name`, `location`, and optional `project` (to be made non-optional i/when the `project` field becomes available via the `enumerateSnapshots` endpoint). Instances of this class will be nested within `TDRSourceRef` as an attribute called `snapshot`, creating a redundancy between their `id` attributes. We'll need to assert that the ID's match. The assertion that the projects match moves to the plugin and is only performed if the `TDRSnapshot` instance's project is not `None`.

All `TDRClient` methods return instances of `TDRSnapshot`.

File: src/azul/terra.py [hunk 1]
```diff
@@ -405,38 +407,42 @@ def _insufficient_access(self, resource: str) -> Exception:
...
-        )
-        return self.TDRSource(project=source['dataProject'],
-                              id=source['id'],
-                              location=storage['region'])
+    def get_snapshot(self, source_spec: TDRSourceSpec) -> TDRSnapshot:
+        snapshot = self._get_snapshot(source_spec)
+        id = snapshot['id']
+        name = snapshot['name']
+        require(source_spec.name == name,
+                f'Sources {source_spec.name!r} and {name!r} both have ID {id}')
```
Comment (amarjandu) [inline src/azul/terra.py:431]: A little confused by this comment, how do we know that the `id` is the same?

**Review by amarjandu — CHANGES_REQUESTED** [review]
Just some small nits.

File: src/azul/terra.py (hunk 1 above)
Comment (noah-aviel-dove) [reply src/azul/terra.py:431]: I made myself very confused while trying to figure out whether this message was accurate or not and decided that either way it was just poorly worded.

File: test/service/test_repository_proxy.py [hunk 2]
```diff
@@ -173,23 +179,37 @@ def test_repository_files_proxy(self, mock_get_cached_sources):
...
+                        #       https://github.com/HumanCellAtlas/dcp2/issues/54
+                        project=None)
             for i, source_name in enumerate(self.mock_source_names + extra_sources)
         }
         mock_source_jsons = [
             {
-                'id': id,
-                'spec': str(TDRSourceSpec.parse(self.make_mock_source_spec(name)).effective)
+                'id': snapshot.id,
+                'spec': str(TDRSourceSpec.parse(self.make_mock_source_spec(snapshot.name)).effective),
```
Comment (amarjandu) [inline test/service/test_repository_proxy.py:202]: Ah another one. 87 char.

**Review by amarjandu — APPROVED** [review]
good call adding the nested method in `from_json`

**Review by amarjandu — DISMISSED** [review]

**Comment by hannes-ucsc** [comment]
Closing this for now so that it doesn't show up in my queue. I am code owner so I can't cancel the review request.

**Pull request closed without merging**